### PR TITLE
WIP and help wanted Elasticsearch & Memcached templates

### PR DESCRIPTION
**Pull request by shicholas**
I realize that I'm a n00b when it comes to this devops stuff, but I have been having trouble following along the ``Development.md`` docs to test templates out on this branch. But reading the dev document, I learned that I could still submit WIP PRs.

Somethings that confused me, and that I still might be blatantly wrong about:

* ``convox install`` is not the equivalent to creating a rack. The first command listed, ``convox install --stack-name dev`` made me believe I was creating a test rack to test changes/develop on. My understanding that because I already have pre-existing convox racks, I do not need to run that command at all.
* instead of a ``git clone`` pulling the package with ``go`` and working within ``$GOPATH/src/github.com/convox/rack`` is better b/c it updates the binstub at ``$GOPATH/bin/convox`` when building the project
* running ``WEB_PID=$(convox api get /apps/$STACK_NAME/processes | jq -r '.[] | select(.name == "web") | .id' | head -1)`` is necessary for the local Rack environment to know about AWS secrets
* "Rack" refers to the project name whereas a rack refers to servers that hold a group of applications.

So I have these two templates, but I have no idea if they work or not. I would like to figure out how to create a service that interfaces with my AWS account, I tried following along https://github.com/convox/rack/pull/622#issuecomment-218513094 but like the author of that PR had trouble getting this up and running.

Let me know how I can help!

### CODE REVIEW

File: Development.md [hunk 1]
```diff
@@ -91,9 +97,11 @@ PASSWORD=45e0f109-3f56-4b30-9b5a-b0939b8a4c25
 ...
 ```
 
+These secrets can be used in your local rack docker vm for interfacing with AWS.
```
Comment (shicholas) [inline Development.md]: no clue if this is true or not. I don't know and couldn't find out how this info gets passed into ``convox start``.

File: Development.md [hunk 2]
```diff
@@ -78,7 +84,7 @@ $ convox exec $WEB_PID env --app $STACK_NAME > .env
 Now you have a bunch of secrets that will let you interact with AWS APIs from your laptop:
 
 ```
-$ cat .env
+$ cat $GOPATH/src/github.com/convox/rack
```
Comment (shicholas) [inline Development.md]: is this the right place for this file?

File: Development.md [hunk 3]
```diff
@@ -91,9 +97,11 @@ PASSWORD=45e0f109-3f56-4b30-9b5a-b0939b8a4c25
 ...
 ```
 
+These secrets can be used in your local rack docker vm for interfacing with AWS.
+
 ## Local Rack Docker VM
 
-A local Rack is started with `convox start` which requires a working Docker environment. To setup a Docker environment, see the [Docker Machine](https://docs.docker.com/machine/) docs. You can then run the project:
+A local rack is started with `convox start` which requires a working Docker machine environment. To setup a Docker environment by creating a Docker machine, see the [Docker Machine](https://docs.docker.com/machine/) docs. You can then run the project:
```
Comment (shicholas) [inline Development.md]: here are cases where the "Rack" v. "rack" confused me.

And also fwiw I think a "Docker Machine" environment is unique to a regular "Docker environment" b/c I view the latter as just running one-off containers.

File: api/models/templates/service/elasticsearch.tmpl [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    "Resources": {
+      "Domain": {
+        "Type": "AWS::Elasticsearch::Domain",
+        "Properties": {
+          "AccessPolicies": {
+            "Version": "2016-05-15",
+            "Statement": [
+               {
+                 "Effect": "Allow",
+                 "Principal": {"AWS": "arn:aws:iam::123456789012:root" },
```
Comment (shicholas) [inline api/models/templates/service/elasticsearch.tmpl]: obviously wrong :). but hopefully I can figure out how to test this soon.

**Comment by nzoschke** [comment]
Hey thanks for the initial work and questions!

> My understanding that because I already have pre-existing convox racks, I do not need to run that command at all.

That is correct. If you have one rack installed and running, you can develop "against" that. The docs try to address that here:

>> You can also use any existing Rack with the caveat that running a local Rack against it could have side effects like terminating instances.

> instead of a git clone pulling the package with go

That's right. Rack is a Go project, so a working GOPATH and using the `go` tool is the best way to work with the project.

> running ... is necessary for the local Rack environment to know about AWS secrets

That's right, and this snippet is surely some wizardry.

On AWS the rack `web` process is running with a bunch of AWS secrets and resource names in its environment. You can introspect any running process with the `convox exec` command and its process id, so this snippet figures out the process id and runs `env`.

Then you can save this environment to your laptop and start a `web` binary in development and to make rapid changes locally.

Finally you tell the `convox` CLI to talk to your local web server.

> "Rack" refers to the project name whereas a rack

Naming is hard! I say the project name is `convox/rack`, following GitHub and Go conventions. 

"Rack" is also a code name we use to refer to your AWS VPC, instances and other related resources. Sometimes I refer to this same concept as your "VPC" or "Cluster".

I hope that helps a bit. Any feedback on how the Developing.md can be more clear is appreciated.

I now owe it to you and #622 to better explain how to test building new services locally. Aurora, ElasticSearch and Memcached would be **amazing**

**Comment by nzoschke** [comment]
The way to develop new services locally is:

* Go through the [Development Guide](https://github.com/convox/rack/blob/master/Development.md) to `convox start` a local Rack API and `convox login localhost` to get the CLI to talk to it
* Fork one of the files in `api/models/templates/service`, i.e. `cp redis.tmpl memcache.tmpl`

Then you can iterate around:

* Change parts of the CloudFormation template to provision your new service.
* Run `make templates` to compile your template and reload the development rack API
* Run `convox services create memcache`
* Use the CloudFormation Management Console and events to see errors in creating the service

**Comment by nzoschke** [comment]
cc @jagtesh, this same procedure applies to the Aurora PR #622

**Comment by shicholas** [comment]
sorry for the delay, I'm following your instructions but am getting stuck after compiling the templates, I created #804 for memcache and will create a new PR for ES. not sure if y'all wanted to close this or not since https://github.com/convox/rack/pull/670#issuecomment-223642151 was super helpful.

**Comment by ddollar** [comment]
Closing in favor of #804, thank you!

File: api/models/templates/service/elasticsearch.tmpl (hunk 4 above)
Comment (slajax) [inline api/models/templates/service/elasticsearch.tmpl]: @shicholas did you ever get any further with this? I need ES so I was going to start looking into setting this up but it seems like you got pretty far with it here.

**Pull request closed without merging**